### PR TITLE
Implemented Big & Small dripleaf

### PR DESCRIPTION
**Pull request by ipad54**
## Introduction
Implemented Big & Small dripleaf

## Backwards compatibility
BC Break change is in RuntimeEnumDescriber interface, but i think it's not a part of API?

## Tests
https://www.youtube.com/watch?v=n94ewbNj_xw

### CODE REVIEW

**Comment by dktapps** [comment]
> BC Break change is in RuntimeEnumDescriber interface, but i think it's not a part of API?

Yeah, this is a design flaw I didn't think of when I initially wrote that code, but users aren't likely to implement it anyway. It should probably be explicitly marked @internal though.

File: src/block/BigDripleaf.php [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+		$w->bool($this->head);
+	}
+
+	/** @return $this */
+	public function setTilt(DripleafTiltType $tilt) : self{
+		$this->tilt = $tilt;
+		return $this;
+	}
+
+	public function getTilt() : DripleafTiltType{
```
Comment (dktapps) [inline src/block/BigDripleaf.php]: nit: getters are usually placed before setters

File: src/block/BigDripleaf.php [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		return $this->tilt;
+	}
+
+	/** @return $this */
+	public function setHead(bool $head) : self{
+		$this->head = $head;
+		return $this;
+	}
+
+	public function isHead() : bool{
```
Comment (dktapps) [inline src/block/BigDripleaf.php]: same again here

File: src/block/BigDripleaf.php [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+		return
+			$block->hasSameTypeId($this) ||
+			$block->getTypeId() === BlockTypeIds::CLAY ||
+			$block->hasTypeTag(BlockTypeTags::DIRT) ||
+			$block->hasTypeTag(BlockTypeTags::MUD);
+	}
+
+	public function onNearbyBlockChange() : void{
+		if(
+			!$this->head && !$this->getSide(Facing::UP)->hasSameTypeId($this) ||
```
Comment (dktapps) [inline src/block/BigDripleaf.php]: Potential operator precedence issue here, I assume the below is the correct solution?
```suggestion
			(!$this->head && !$this->getSide(Facing::UP)->hasSameTypeId($this)) ||
```

File: src/block/BigDripleaf.php [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		if(
+			!$world->isInWorld($up->getFloorX(), $up->getFloorY(), $up->getFloorZ()) ||
+			$world->getBlock($up)->getTypeId() !== BlockTypeIds::AIR
+		){
+			return false;
+		}
+
+		$tx = new BlockTransaction($world);
+
+		$tx->addBlock($pos, (clone $head)->setHead(false));
```
Comment (dktapps) [inline src/block/BigDripleaf.php]: what about the tilt?

I'm getting the impression that BIG_DRIPLEAF_STEM ought to be separated from BIG_DRIPLEAF, since the tilt property does not apply to the stem.

File: src/block/BigDripleaf.php [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+		$tx->addBlock($up, VanillaBlocks::BIG_DRIPLEAF()
+			->setFacing($head->getFacing())
+			->setHead(true)
+		);
```
Comment (dktapps) [inline src/block/BigDripleaf.php]: Again, looks like the tilt has not been considered here

File: src/block/BigDripleaf.php [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+		if(!$this->head || $this->tilt->equals(DripleafTiltType::FULL())){
+			return [];
+		}
+		return [
+			AxisAlignedBB::one()
+			->trim(Facing::DOWN, 11 / 16)
+			->trim(Facing::UP, match($this->tilt->id()) {
+				DripleafTiltType::NONE()->id(), DripleafTiltType::UNSTABLE()->id() => 1 / 16,
+				DripleafTiltType::PARTIAL()->id() => 3 / 16,
+				default => throw new AssumptionFailedError("All types should be covered")
```
Comment (dktapps) [inline src/block/BigDripleaf.php]: This block below `AxisAlignedBB::one()` needs to be indented by 1

File: src/block/BigDripleaf.php [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+			})
+		];
+	}
+
+	public function isFlammable() : bool{
+		return true;
+	}
```
Comment (dktapps) [inline src/block/BigDripleaf.php]: The flame encouragement and flammability values should be set instead of overriding this - this is not supposed to be overridden

A Bedrock data dump suggests that the flammability is `100` and the flame encouragement is `15`

File: src/block/SmallDripleaf.php [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+use pocketmine\math\Facing;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\BlockTransaction;
+use function mt_rand;
+
+class SmallDripleaf extends Transparent{
+	use HorizontalFacingTrait;
+
+	protected bool $upperBlock = false;
```
Comment (dktapps) [inline src/block/SmallDripleaf.php]: `top` would be the idiomatic way of naming this, for consistency with other blocks such as double plants and doors

Comment (dktapps) [reply src/block/SmallDripleaf.php]: Perhaps this class can use `DoublePlant` as a base? Not an instruction, but a suggestion

File: src/block/SmallDripleaf.php [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+			$pos = $bottomBlock->getSide(Facing::UP, $i)->getPosition();
+			if(!$world->isInWorld($pos->getFloorX(), $pos->getFloorY(), $pos->getFloorZ())){
+				break;
+			}
+			$replace = $world->getBlock($pos);
+			if(!$replace->hasSameTypeId($this) && $replace->getTypeId() !== BlockTypeIds::AIR){
+				break;
+			}
+			$tx->addBlock($pos, VanillaBlocks::BIG_DRIPLEAF()->setFacing($this->facing));
+			$growed++;
```
Comment (dktapps) [inline src/block/SmallDripleaf.php]: `grown` would be more idiomatic

File: src/block/SmallDripleaf.php [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		$bottomBlock = $this->upperBlock ? $this->getSide(Facing::DOWN) : $this;
+		if(!$this->hasSameTypeId($bottomBlock)){
+			return false;
+		}
+		$top = $bottomBlock->getPosition();
+		$world = $top->getWorld();
+		$tx = new BlockTransaction($world);
+		$growed = 0;
+
+		for($i = 0; $i < mt_rand(2, 5); $i++){
```
Comment (dktapps) [inline src/block/SmallDripleaf.php]: This is going to re-run `mt_rand()` for every loop iteration - probably not what you wanted

File: src/block/SmallDripleaf.php [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+			}
+			$tx->addBlock($pos, VanillaBlocks::BIG_DRIPLEAF()->setFacing($this->facing));
+			$growed++;
+			$top = $pos;
+		}
+
+		if($growed > 1){
+			$block = $tx->fetchBlock($top);
+			if($block instanceof BigDripleaf && $block->getTypeId() === BlockTypeIds::BIG_DRIPLEAF){
+				$block->setHead(true);
```
Comment (dktapps) [inline src/block/SmallDripleaf.php]: This logic feels a bit weird

File: src/data/bedrock/block/convert/BlockStateReader.php [hunk 12]
```diff
@@ -349,6 +350,17 @@ public function readWallConnectionType(string $name) : ?WallConnectionType{
...
+	/** @throws BlockStateDeserializeException */
+	public function readDripleafTiltType() : DripleafTiltType{
+		return match($type = $this->readString(BlockStateNames::BIG_DRIPLEAF_TILT)){
+			StringValues::BIG_DRIPLEAF_TILT_NONE => DripleafTiltType::NONE(),
+			StringValues::BIG_DRIPLEAF_TILT_UNSTABLE => DripleafTiltType::UNSTABLE(),
+			StringValues::BIG_DRIPLEAF_TILT_PARTIAL_TILT => DripleafTiltType::PARTIAL(),
+			StringValues::BIG_DRIPLEAF_TILT_FULL_TILT => DripleafTiltType::FULL(),
+			default => throw $this->badValueException(BlockStateNames::BIG_DRIPLEAF_TILT, $type),
+		};
+	}
```
Comment (dktapps) [inline src/data/bedrock/block/convert/BlockStateReader.php]: I'm not really clear what we're gaining by putting this here, as opposed to in the `BlockStateToObjectDeserializer` directly. It's only used once.

File: src/data/bedrock/block/convert/BlockStateWriter.php [hunk 13]
```diff
@@ -318,6 +319,18 @@ public function writeWallConnectionType(string $name, ?WallConnectionType $wallC
...
+	public function writeDripleafTiltType(DripleafTiltType $tiltType) : self{
+		$this->writeString(BlockStateNames::BIG_DRIPLEAF_TILT, match($tiltType->id()){
+			DripleafTiltType::NONE()->id() => StringValues::BIG_DRIPLEAF_TILT_NONE,
+			DripleafTiltType::UNSTABLE()->id() => StringValues::BIG_DRIPLEAF_TILT_UNSTABLE,
+			DripleafTiltType::PARTIAL()->id() => StringValues::BIG_DRIPLEAF_TILT_PARTIAL_TILT,
+			DripleafTiltType::FULL()->id() => StringValues::BIG_DRIPLEAF_TILT_FULL_TILT,
+			default => throw new BlockStateSerializeException("Invalid Dripleaf tilt type " . $tiltType->name())
+		});
+		return $this;
+	}
```
Comment (dktapps) [inline src/data/bedrock/block/convert/BlockStateWriter.php]: Same again here.

**Review by dktapps — CHANGES_REQUESTED** [review]
I haven't yet tested this, but I have a few thoughts about the design of this PR.

Most notably, I think that `BIG_DRIPLEAF` needs to be split into stem vs. head parts, because the head has properties (tilt) which are useless to the stems.

File: src/block/SmallDripleaf.php (hunk 8 above)
Comment (ipad54) [reply src/block/SmallDripleaf.php]: Double Plants are flowable blocks but small dripleaf is not

> Perhaps this class can use `DoublePlant` as a base? Not an instruction, but a suggestion

File: src/block/BaseBigDripleaf.php [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+			!$world->isInWorld($up->getFloorX(), $up->getFloorY(), $up->getFloorZ()) ||
+			$world->getBlock($up)->getTypeId() !== BlockTypeIds::AIR
+		){
+			return false;
+		}
+
+		$tx = new BlockTransaction($world);
+
+		$tx->addBlock($pos, VanillaBlocks::BIG_DRIPLEAF_STEM()->setFacing($head->getFacing()));
+		$tx->addBlock($up, VanillaBlocks::BIG_DRIPLEAF()->setFacing($head->getFacing()));
```
Comment (dktapps) [inline src/block/BaseBigDripleaf.php]: what about the tilt?

File: src/block/BaseBigDripleaf.php [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+abstract class BaseBigDripleaf extends Transparent{
+	use HorizontalFacingTrait;
+
+	abstract protected function isHead() : bool;
+
+	private function canBeSupportedBy(Block $block) : bool{
+		//TODO: Moss block
+		return
+			$block instanceof BaseBigDripleaf ||
```
Comment (dktapps) [inline src/block/BaseBigDripleaf.php]: I don't think this is correct. Both stems and heads should only be able to be supported by other stems. A dripleaf head should not be able to provide support to more dripleaf.

File: src/block/BaseBigDripleaf.php (hunk 14 above)
Comment (ipad54) [reply src/block/BaseBigDripleaf.php]: In vanilla tilt restores to NONE when block grows

File: src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php [hunk 16]
```diff
@@ -784,6 +786,26 @@ private function registerDeserializers() : void{
...
 				->setAttachmentType($in->readBellAttachmentType());
 		});
+		$this->map(Ids::BIG_DRIPLEAF, function(Reader $in) : Block{
+			$block =
+				($head = $in->readBool(StateNames::BIG_DRIPLEAF_HEAD)) ?
+					Blocks::BIG_DRIPLEAF() :
+					Blocks::BIG_DRIPLEAF_STEM();
+			$block->setFacing($in->readLegacyHorizontalFacing());
+			if($head){
+				/** @var BigDripleaf $block */
```
Comment (dktapps) [inline src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php]: This shouldn't be needed if the code is properly written.

File: src/block/BaseBigDripleaf.php (hunk 14 above)
Comment (dktapps) [reply src/block/BaseBigDripleaf.php:114]: honestly this sounds like a bug. have you tested it in java?

Comment (dktapps) [reply src/block/BaseBigDripleaf.php]: honestly this sounds like a bug. have you tested in Java?

Comment (ipad54) [reply src/block/BaseBigDripleaf.php]: Tested it on Java a minute ago, Java seems to be also work this way.

https://www.youtube.com/watch?v=3tPl22WUuXE

Comment (dktapps) [reply src/block/BaseBigDripleaf.php]: Fair enough, I guess this can be chalked up to another Mojang inconsistency.

File: src/block/SmallDripleaf.php (hunk 8 above)
Comment (dktapps) [reply src/block/SmallDripleaf.php]: > Double Plants are flowable blocks but small dripleaf is not

Right, I didn't think of that.

File: src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php (hunk 16 above)
Comment (ipad54) [reply src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php]: PHPStan says ```Call to an undefined method pocketmine\block\BigDripleaf|pocketmine\block\BigDripleafStem::setTilt(). ```

File: src/block/BaseBigDripleaf.php [hunk 17]
```diff
@@ -38,10 +38,10 @@ abstract class BaseBigDripleaf extends Transparent{
 
 	abstract protected function isHead() : bool;
 
-	private function canBeSupportedBy(Block $block) : bool{
+	private function canBeSupportedBy(Block $block, bool $head = false) : bool{
```
Comment (dktapps) [inline src/block/BaseBigDripleaf.php]: Avoid optional parameters.

File: src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php (hunk 16 above)
Comment (dktapps) [reply src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php]: As I said, the code is improperly written.

File: src/block/BaseBigDripleaf.php [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+abstract class BaseBigDripleaf extends Transparent{
+	use HorizontalFacingTrait;
+
+	abstract protected function isHead() : bool;
+
+	private function canBeSupportedBy(Block $block, bool $head) : bool{
+		//TODO: Moss block
+		return
+			$block instanceof BaseBigDripleaf && $block->isHead() === $head ||
```
Comment (dktapps) [inline src/block/BaseBigDripleaf.php]: Precedence issue here
```suggestion
			($block instanceof BaseBigDripleaf && $block->isHead() === $head) ||
```

File: src/world/format/io/data/BedrockWorldData.php [hunk 19]
```diff
@@ -155,18 +154,12 @@ protected function load() : CompoundTag{
...
-		}
 		if($version > self::CURRENT_STORAGE_VERSION){
 			throw new UnsupportedWorldFormatException("LevelDB world format version $version is currently unsupported");
 		}
 		//StorageVersion is rarely updated - instead, the game relies on the NetworkVersion tag, which is synced with
 		//the network protocol version for that version.
 		$protocolVersion = $worldData->getInt(self::TAG_NETWORK_VERSION, Limits::INT32_MAX);
-		if($protocolVersion === Limits::INT32_MAX){
-			throw new CorruptedWorldException(sprintf("Missing '%s' tag in level.dat", self::TAG_NETWORK_VERSION));
-		}
```
Comment (dktapps) [inline src/world/format/io/data/BedrockWorldData.php]: Why is this changed?

Comment (ipad54) [reply src/world/format/io/data/BedrockWorldData.php]: I didn't change it, maybe it's a github bug

Comment (dktapps) [reply src/world/format/io/data/BedrockWorldData.php]: Diff suggests it came from this commit: https://github.com/pmmp/PocketMine-MP/pull/5835/commits/8d7023f9ad7ac9c72b09b55bb09cfce4f33d2ecf

Comment (ipad54) [reply src/world/format/io/data/BedrockWorldData.php]: Should be fixed now

**Review by dktapps — APPROVED** [review]
